x/lightclient/keeper/hook_listener.go

```diff
@@ -0,0 +1,120 @@
+package keeper
+
+import (
+	"errors"
+	"time"
+
+	tmprotocrypto "github.com/cometbft/cometbft/proto/tendermint/crypto"
+	cryptocodec "github.com/cosmos/cosmos-sdk/crypto/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	ibcclienttypes "github.com/cosmos/ibc-go/v7/modules/core/02-client/types"
+	ibctm "github.com/cosmos/ibc-go/v7/modules/light-clients/07-tendermint"
+	"github.com/dymensionxyz/dymension/v3/x/lightclient/types"
+
+	rollapptypes "github.com/dymensionxyz/dymension/v3/x/rollapp/types"
+)
+
+var _ rollapptypes.RollappHooks = rollappHook{}
+
+// Hooks wrapper struct for rollapp keeper.
+type rollappHook struct {
+	rollapptypes.StubRollappCreatedHooks
+	k Keeper
+}
+
+// RollappHooks returns the wrapper struct.
+func (k Keeper) RollappHooks() rollapptypes.RollappHooks {
+	return rollappHook{k: k}
+}
+
+// AfterUpdateState is called after a state update is made to a rollapp.
+// This hook checks if the rollapp has a canonical IBC light client and if the Consensus state is compatible with the state update
+// and punishes the sequencer if it is not
+func (hook rollappHook) AfterUpdateState(
+	ctx sdk.Context,
+	rollappId string,
+	stateInfo *rollapptypes.StateInfo,
+	isFirstStateUpdate bool,
+	previousStateHasTimestamp bool,
+) error {
+	canonicalClient, found := hook.k.GetCanonicalClient(ctx, rollappId)
+	if !found {
+		canonicalClient = hook.k.GetProspectiveCanonicalClient(ctx, rollappId, stateInfo)
+		if canonicalClient != "" {
+			hook.k.SetCanonicalClient(ctx, rollappId, canonicalClient)
+		}
+		return nil
+	}
+	bds := stateInfo.GetBDs()
+	for i, bd := range bds.GetBD() {
+		// Check if any optimistic updates were made for the given height
+		tmHeaderSigner, found := hook.k.GetConsensusStateSigner(ctx, canonicalClient, bd.GetHeight())
+		if !found {
+			continue
+		}
```

Comment:
it's more straightforward to iterate over all BDs  and compare to any consensus states at `h-1`